pil-analyzer/src/pil_analyzer.rs

```diff
@@ -361,7 +361,7 @@ impl PILAnalyzer {
     /// Creates and returns a map for every referenced trait and every concrete type to the
     /// corresponding trait implementation function.
     fn resolve_trait_impls(&mut self) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
-        let mut trait_solver = TraitsResolver::new(&self.implementations);
+        let mut trait_solver = TraitsResolver::new(&self.implementations, &self.definitions);
 
         let mut resolve_references = |expr: &Expression| {
```

Comment:
I think all the following code in this function should be simplified to this:
```rust
        // Now start resolving at the "entry points", i.e. all proof items
        // and all columns.
        // Other symbols might be unused, so we do not need to resolve them.
        let entry_points = self
            .definitions
            .values()
            // TODO simplify?
            .filter_map(|(symbol, value)| (symbol.kind == SymbolKind::Other()).then(|| value))
            .flatten()
            .flat_map(|fv| fv.children())
            .chain(&self.proof_items);
        for expr in entry_points {
            trait_solver.resolve_trait_references(&expr)?;
        }
        Ok(trait_solver.solved_impls())
```

The idea here is the following: We supply the trait resolver with all the "entry points" into the program. It follows the call graph and resolves all references to trait functions it finds. This has the advantage that we do not resolve functions (for example in the stdlib) that are never called.

Since we need to process the call graph anyway, I think this is a big simplification.